File: src/VoidRay/Services/NetTools.cs
using System.Diagnostics;
using System.Net;
using System.Net.Sockets;
using System.Text.Json;

namespace VoidRay.Services;

public static class NetTools
{
    /// <summary>TCP handshake latency to the server, in ms (null on failure).</summary>
    public static async Task<int?> TcpPingAsync(string host, int port, int timeoutMs = 3000)
    {
        try
        {
            using var cts = new CancellationTokenSource(timeoutMs);
            var addresses = await Dns.GetHostAddressesAsync(host, cts.Token);
            var address = addresses.FirstOrDefault(a => a.AddressFamily == AddressFamily.InterNetwork) ?? addresses.First();
            using var client = new TcpClient(address.AddressFamily);
            var sw = Stopwatch.StartNew();
            await client.ConnectAsync(address, port, cts.Token);
            return (int)Math.Max(1, sw.ElapsedMilliseconds);
        }
        catch
        {
            return null;
        }
    }

    public sealed record ExitInfo(string Ip, string? Country, string? CountryCode, int DelayMs);

    /// <summary>Queries the public IP through the local proxy: proves the tunnel works end to end.</summary>
    public static async Task<ExitInfo?> CheckExitAsync(int httpPort, CancellationToken ct = default)
    {
        using var http = new HttpClient(new HttpClientHandler
        {
            Proxy = new WebProxy($"http://127.0.0.1:{httpPort}"),
            UseProxy = true,
        })
        { Timeout = TimeSpan.FromSeconds(10) };
        http.DefaultRequestHeaders.UserAgent.ParseAdd("VoidRay/1.0");

        try
        {
            var sw = Stopwatch.StartNew();
            var json = await http.GetStringAsync("https://ipwho.is/?fields=ip,country,country_code", ct);
            var delay = (int)sw.ElapsedMilliseconds;
            using var doc = JsonDocument.Parse(json);
            var r = doc.RootElement;
            return new ExitInfo(
                r.GetProperty("ip").GetString() ?? "?",
                r.TryGetProperty("country", out var c) ? c.GetString() : null,
                r.TryGetProperty("country_code", out var cc) ? cc.GetString() : null,
                delay);
        }
        catch
        {
            try
            {
                var sw = Stopwatch.StartNew();
                var ip = (await http.GetStringAsync("https://api.ipify.org", ct)).Trim();
                return new ExitInfo(ip, null, null, (int)sw.ElapsedMilliseconds);
            }
            catch
            {
                return null;
            }
        }
    }

    public static int FindFreePort(int preferred)
    {
        try
        {
            var probe = new TcpListener(IPAddress.Loopback, preferred);
            probe.Start();
            probe.Stop();
            return preferred;
        }
        catch (SocketException)
        {
            var any = new TcpListener(IPAddress.Loopback, 0);
            any.Start();
            var port = ((IPEndPoint)any.LocalEndpoint).Port;
            any.Stop();
            return port;
        }
    }
}

File: android/app/src/main/java/win/enderr/voidray/vpn/NetTools.kt
package win.enderr.voidray.vpn

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Proxy
import java.net.Socket
import java.net.URL

object NetTools {
    /** TCP handshake latency to the server in ms (null on failure). */
    fun tcpPing(host: String, port: Int, timeoutMs: Int = 3000): Int? = try {
        val address = InetAddress.getAllByName(host).let { all -> all.firstOrNull { it.address.size == 4 } ?: all.first() }
        Socket().use { s ->
            val start = System.nanoTime()
            s.connect(InetSocketAddress(address, port), timeoutMs)
            maxOf(1, ((System.nanoTime() - start) / 1_000_000).toInt())
        }
    } catch (e: Exception) {
        null
    }

    class ExitInfo(val ip: String, val country: String?, val countryCode: String?, val delayMs: Int)

    /**
     * Queries the public IP through Xray's local SOCKS inbound. The app itself is
     * excluded from the VPN, so this is how it proves the tunnel works end to end.
     */
    fun checkExit(socksPort: Int): ExitInfo? {
        val proxy = Proxy(Proxy.Type.SOCKS, InetSocketAddress("127.0.0.1", socksPort))
        fun get(url: String): Pair<String, Int> {
            val conn = URL(url).openConnection(proxy) as HttpURLConnection
            try {
                conn.connectTimeout = 10_000
                conn.readTimeout = 10_000
                conn.setRequestProperty("User-Agent", "VoidRay-Android/1.0")
                val start = System.nanoTime()
                val body = conn.inputStream.use { it.readBytes().toString(Charsets.UTF_8) }
                return body to ((System.nanoTime() - start) / 1_000_000).toInt()
            } finally {
                conn.disconnect()
            }
        }
        return try {
            val (body, delay) = get("https://ipwho.is/?fields=ip,country,country_code")
            val o = JSONObject(body)
            ExitInfo(o.getString("ip"), o.optString("country").ifEmpty { null }, o.optString("country_code").ifEmpty { null }, delay)
        } catch (e: Exception) {
            try {
                val (ip, delay) = get("https://api.ipify.org")
                ExitInfo(ip.trim(), null, null, delay)
            } catch (e2: Exception) {
                null
            }
        }
    }
}
